README.md

<p align="center">
  <img width="300" src="https://github.com/dinko-pehar/mock_data/blob/master/assets/mock_data_logo.png" alt="Mock Data Logo" /><br>
  <a href="https://github.com/dinko-pehar/mock_data/actions"><img src="https://github.com/dinko-pehar/mock_data/workflows/Dart%20CI/badge.svg?branch=master" alt="Build Status"></a>
  <a href="LICENSE.txt"><img src="https://img.shields.io/github/license/dinkopehar/mock_data.svg" alt="License"></a>
  <a href="https://pub.dev/packages/mock_data"><img src="https://img.shields.io/pub/v/mock_data" alt="Version"></a><br>
  Generate random data using <img src="https://raw.githubusercontent.com/github/explore/80688e429a7d4ef2fca1e82350fe8e3517d3494d/topics/dart/dart.png" alt="Dart" width="16" height="16">
</p>

## Features

API provides generation of:

- *Integers* in any range of numbers
- *Strings* with various characters and length
- *Colors* represented with needed color space
- *Date*s between different moments in time
- *Name*s such as John or Mary
- *UUIDv4* and *Timestamp-first UUIDs*
- *URL*s with many fragments
- *IPv4* and *IPv6*
- *Locations* as either random points or in specified radius

## Usage

A simple usage example:

```dart
import 'package:mock_data/mock_data.dart';

main() {

  mockName();           // Generate male or female first name. 
  mockName('male');     // Generate male first name. 
  mockName('female');   // Generate female first name. 

  mockInteger(1, 6);    // Generate integer in range from 1 do 6.

  mockString(16);       // Generate string of length 16.
  
  mockIPv4();           // Generate IPv4 represented with 
                        // format(default is '*.*.*.*') as String. 
  
  mockIPv6();           // Generate IPv6, same usage as with IPv4.

  mockColor('hex');     // Generate color represented in hex format.
  mockColor('rgb');     // Generate color represented in RGB format.
  
  mockUUID();           // Generate UUIDv4
  
  mockLocation();       // Generate random location.
  mockLocation(53.55, 108.14, 1000); // Generate random location in 1000 m radius from the specified coordinates.
  
}
```

These are some basic examples. There are many more methods and they all support tweeking of parameters to suit you in generating random data. By reading [examples][] you can learn more about functionality and usage of `mock_data`

### Contributors
- [Dinko Pehar](https://github.com/dinko-pehar)
- [Nagasai Bodepudi](https://github.com/nagasaiBodepudi)
- [Volodymyr Buberenko](https://github.com/vbuberen)
- [Ivan Murashov](https://github.com/kifio)

### License
[MIT][]

[MIT]: https://github.com/dinkopehar/mock_data/blob/master/LICENSE.txt
[examples]: https://pub.dev/packages/mock_data#-example-tab-


CHANGELOG.md
## 2.0.1
- UPDATED: Bump Dart to version 3

## 2.0.0
- NEW: 'mockLocation' for location mocking
- UPDATED: Migrated to sound null-safety.
- IMPROVED: Minor tests refactoring

## 1.2.7
- IMPROVED: README.md section reduced to basic examples. For more advanced
            usage refer to examples section.
- REMOVED: TODO file. Refer to Project tab for Roadmap.
- UPDATED: Issue templates when opening new issue.

## 1.2.6

- NEW: `mockUUID` function for generating version 4 UUID.
- IMPROVED: Health suggestions for analysis tab.
- CHANGED: TravisCI replaced with GitHub Actions.

## 1.2.2 - 1.2.5

- [Maintenance](https://pub.dartlang.org/packages/mock_data#-analysis-tab-) score increased

## 1.2.1

- UPDATED: Added support for IPv6 instead of IPv4 only.
- UPDATED: Added new color spaces for `mockColor`
- UPDATED: Added new male and female names
- IMPROVED: Added more tests
- CHANGED: `mockName` defaults to _'any'_ instead of empty string
- FIX: `mockName` test for gender changed to more valid one

## 1.2.0

- NEW: `mockColor`, `mockUrl` and `mockDate` functions

## 1.1.0

- NEW: `mockName` function
- Improved documentation
- FIX: Removed last dot that was generated using _mockIPv4_ function

## 1.0.1

- FIX: Added proper documentation link

## 1.0.0

- Initial version


lib/mock_data.dart
/// Generate random data using Dart.
///
/// List of random data to generate:
/// * **colors** with various color spaces
/// * **date** between different moments in time
/// * **integers** in between range of numbers
/// * **IPs** ~ IPv4 and IPv6
/// * **name**; male or female first name
/// * **strings** of different length and characters
/// * **url** with different parts; routes, GET query parameters and fragments
/// * **uuid** ~ version 4, timestamp-first and null
///
/// Inspired by: https://www.npmjs.com/package/mock-data

library mock_data;

export 'src/mock_color.dart';
export 'src/mock_date.dart';
export 'src/mock_integer.dart';
export 'src/mock_ip.dart';
export 'src/mock_location.dart';
export 'src/mock_name.dart';
export 'src/mock_string.dart';
export 'src/mock_url.dart';
export 'src/mock_uuid.dart';

/// Generic function to generate list of mocks.
///
/// [mockFunction] represents a function from _mock_data_ library.
/// It is passed as first class object (without invocation).
///
/// [numberOfMocks] represents number of invocations of [mockFunction].
///
/// `mockRange` can also accept all parameters of [mockFunctions] as
/// optional named parameters. Default values for optional parameters
/// is same as for _mock_data_ functions.
///
/// Returns [List<E>].
///
/// Example usage:
/// ```dart
///   // returns list of length 15 containing
///   // random IPs.
///   mockRange(mockIPv4, 15)
///
///   // returns list of length 5 containing
///   // random IPs with starting groups as
///   // 192.168. and last 2 groups random int
///   // in range 0 to 255, inclusive.
///   mockRange(mockIPv4, 5, format:192.168.*.*)
///
///   // similar to IPv4.
///   mockRange(mockIPv6, 5, format:192.168.*.*)
///
///   // returns list of length 5 containing
///   // random integers ranging from -15
///   // to 20, inclusive.
///   mockRange(mockInteger, 3, min: -15, max:20)
///
///   // returns list of length 5 containing
///   // random strings of length 16 which are
///   // consisted of any combination of digits,
///   // lowercase and uppercase characters.
///   mockRange(mockString, 5)
///
///   // returns list of length 15 containing
///   // random males and females first names.
///   mockRange(mockName, 15)
///
///   // returns list of length 7 containing
///   // random colors represented in hex format.
///   mockRange(mockColor, 7, returnModel: 'hex')
///
///   // returns list of length 4 containing
///   // random URLs with query string and permalink.
///   mockRange(mockUrl, 4, withQuery: true, withFragment: true)
///
///   // returns list of length 5 containing
///   // random DateTime objects in between 1970-01-01 01:00:00.000 and
///   // current time(now).
///   mockRange(mockDate, 5, secondMoment: DateTime(2000))
///
///   // returns list of length 3 containing
///   // timestamp-first UUIDs.
///   mockRange(mockUUID, 3, uuidType: 'timestamp-first')
///
///   // returns list containing random locations in 1km radius from Moscow Kremlin.
///   mockRange(mockLocation, 55.7520, 37.6175, 1000)
/// ```
List<E> mockRange<E>(Function mockFunction, int numberOfMocks,
    {int lengthOfMockedString = 16,
    String include = '!',
    int min = 1,
    int max = 10,
    String format = '*.*.*.*',
    String gender = 'any',
    String returnModel = 'rgb',
    String scheme = '*',
    bool withPath = false,
    withQuery = false,
    withFragment = false,
    DateTime? firstMoment,
    DateTime? secondMoment,
    String uuidType = 'ver4',
    double centerLat = 55.7520,
    double centerLon = 37.6175,
    int radius = 1000}) {
  // Get name of passed function.
  var f = RegExp(r"'(.+)'").firstMatch(mockFunction.toString())?.group(1);

  // Call proper function with passed arguments.
  switch (f) {
    case 'mockString':
      return List<E>.generate(
          numberOfMocks, (_) => mockFunction(lengthOfMockedString, include));
    case 'mockInteger':
      return List<E>.generate(numberOfMocks, (_) => mockFunction(min, max));
    case 'mockIPv4':
      return List<E>.generate(numberOfMocks, (_) => mockFunction(format));
    case 'mockIPv6':
      return List<E>.generate(numberOfMocks, (_) => mockFunction(format));
    case 'mockName':
      return List<E>.generate(numberOfMocks, (_) => mockFunction(gender));
    case 'mockColor':
      return List<E>.generate(numberOfMocks, (_) => mockFunction(returnModel));
    case 'mockUrl':
      return List<E>.generate(numberOfMocks,
          (_) => mockFunction(scheme, withPath, withQuery, withFragment));
    case 'mockDate':
      return List<E>.generate(
          numberOfMocks, (_) => mockFunction(firstMoment, secondMoment));
    case 'mockUUID':
      return List<E>.generate(numberOfMocks, (_) => mockFunction(uuidType));
    case 'mockLocation':
      return List<E>.generate(
          numberOfMocks, (_) => mockFunction(centerLat, centerLon, radius));
    default:
      return <E>[]; // Empty List if function is not from mock_data lib.
  }
}
